contrib/initramfs/scripts/zfs.in

```diff
@@ -391,67 +391,36 @@ mount_fs()
 decrypt_fs()
 {
 	local fs="$1"
-
-	# If the 'zfs key' command isn't availible, exit right here.
-	"${ZFS}" 2>&1 | grep -q 'key -l ' || return 0
-
-	# Check if filesystem is encrypted. If not, exit right here.
-	[ "$(get_fs_value "$fs" encryption)" != "off" ] || return 0
-
-	[ "$quiet" != "y" ] && \
-	    zfs_log_begin_msg "Loading crypto wrapper key for $fs"
-
-	# Just make sure that ALL crypto modules module is loaded.
-	# Simplest just to load all...
-	for mod in sun-ccm sun-gcm sun-ctr
-	do
-		[ "$quiet" != "y" ] && zfs_log_progress_msg "${mod} "
-
-		ZFS_CMD="load_module $mod"
-		ZFS_STDERR="$(${ZFS_CMD} 2>&1)"
-		ZFS_ERROR="$?"
-
-		if [ "${ZFS_ERROR}" != 0 ]
-		then
-			[ "$quiet" != "y" ] && zfs_log_failure_msg "${ZFS_ERROR}"
-
-			disable_plymouth
-			echo ""
-			echo "Command: $ZFS_CMD"
-			echo "Message: $ZFS_STDERR"
-			echo "Error: $ZFS_ERROR"
-			echo ""
-			echo "Failed to load $mod module."
-			echo "Please verify that it is availible on the initrd image"
-			echo "(without it it won't be possible to unlock the filesystem)"
-			echo "and rerun:  $ZFS_CMD"
-			/bin/sh
-		else
-			[ "$quiet" != "y" ] && zfs_log_end_msg
+	
+	# If pool encryption is active and the zfs command understands '-o encryption'
+	if [ "$(zpool list -H -o feature@encryption $(echo "${fs}" | awk -F\/ '{print $1}'))" = 'active' ]; then
+
+		# Determine dataset that holds key for root dataset
+		ENCRYPTIONROOT=$(zfs get -H -o value encryptionroot "${fs}")
+		DECRYPT_CMD="${ZFS} load-key ${ENCRYPTIONROOT}"
+
+		# If root dataset is encrypted...
+		if ! [ "${ENCRYPTIONROOT}" = "-" ]; then
+
+			# Prompt with plymouth, if active
+			if [ -e /bin/plymouth ] && /bin/plymouth --ping 2>/dev/null; then
+				plymouth ask-for-password --prompt "Encrypted ZFS password for ${ENCRYPTIONROOT}" \
+					--number-of-tries="3" \
+					--command="${DECRYPT_CMD}"
+
+			#Prompt with systemd, if active 
+			elif [ -e /run/systemd/system ]; then
+				TRY_COUNT=3
+				while [ $TRY_COUNT -gt 0 ]; do
+					systemd-ask-password "Encrypted ZFS password for ${ENCRYPTIONROOT}" --no-tty | \
+						${DECRYPT_CMD} && break
+					TRY_COUNT=$((TRY_COUNT - 1))
+				done
+
+			# Prompt with ZFS tty, otherwise
+			else ${DECRYPT_CMD}
```

Comment:
nit: `${DECRYPT_CMD}` should be on its own line.

Reply:
Done